pilot/pkg/model/push_context.go

```diff
@@ -78,9 +80,115 @@ type PushContext struct {
+// example by deployment, or to deal with very large number of endpoints for a service.
+// We want to avoid passing around large objects - like full list of endpoints for a registry,
+// or the full list of endpoints for a service across registries, since it limits scalability.
+//
+// Future optimizations will include grouping the endpoints by labels, gateway or region to
+// reduce the time when subsetting or split-horizon is used. This desing assumes pilot
+// tracks all endpoints in the mesh and they fit in RAM - so limit is few M endpoints.
+// It is possible to split the endpoint tracking in future.
+type EDSUpdater interface {
+
+	// EDSUpdate is called when the list of endpoints or labels in a ServiceEntry is
+	// changed. For each cluster and hostname, the full list of active endpoints (including empty list)
+	// must be sent. The shard name is used as a key - current implementation is using the registry
+	// name.
+	EDSUpdate(shard, hostname string, entry []*IstioEndpoint) error
+
+	// SvcUpdate is called when a service port mapping definition is updated.
+	// This interface is WIP - labels, annotations and other changes to service may be
+	// updated to force a EDS and CDS recomputation and incremental push, as it doesn't affect
+	// LDS/RDS.
+	SvcUpdate(shard, hostname string, ports map[string]uint32, rports map[uint32]string)
+}
+
+// IstioEndpoint has the information about a single address+port for a specific
+// service and shard. Unlike ServiceInstance, it has less info and require less
+// allocation.
+type IstioEndpoint struct {
+
+	// Labels points to the workload or deployment labels.
+	Labels map[string]string
+
+	// Address is the IP address of the endpoint. Other types may be
+	// supported in future.
+	Address string
+
+	// EndpointPort is the port where the workload is listening, can be different
+	// from the service port.
+	EndpointPort uint32
+
+	// ServicePortName tracks the name of the port, to avoid 'eventual consistency' issues.
+	// Sometimes the Endpoint is visible before Service - so looking up the port number would
+	// fail. Instead the mapping to number is made when the clusters are computed. The lazy
+	// computation will also help with 'on-demand' and 'split horizon' - where it will be skipped
+	// for not used clusters or endpoints behind a gate.
+	ServicePortName string
+
+	// UID identifies the workload, for telemetry purpose.
+	UID string
+
+	// EnvoyEndpoint is a cached LbEndpoint, converted from the data, to
+	// avoid recomputation
+	EnvoyEndpoint *endpoint.LbEndpoint
+
+	// ServiceAccount holds the associated service account.
+	// May be replaced once a better optimization/struct is available for secure naming.
+	ServiceAccount string
+}
+
+// ServiceShards holds the set of endpoint shards of a service. Registries update
+// individual shards incrementally. The shards are aggregated and split into
+// clusters when a push for the specific cluster is needed.
+type ServiceShards struct {
```

Comment:
Not sure how to collapse the 2 data structures. Do you have any other naming proposals ? 
I guess 'ServiceShards' need to change - it's meant to hold the 'shards'/pieces of a service.

Reply:
Well we can just inline the EndpointShard in `ServiceShards`
```
type ServiceShards {
  Shards map[string][]*IstioEndpoint
  ServiceAccounts map[string]bool
}
```
I did code search and only references of EndpointShards is in `eds.go`.

Regarding the naming, maybe `ServiceEndpointShards`? Or `EndpointShards` if we only have one data structure.

Another sort of separate question, why not utilizing existing `mode.Service` data structure and make this part of it? Right now we'll have two DS both are 1:1 mapping to the number of services...

Reply:
Ah, I see what you mean. I didn't inline because the "EndpointShard" will have additional fields - not only IstioEndpoint. One optimization is to group (shard) endpoints by labels, so we don't have to check labels on each IstioEndpoint. There are other grouping/sharding that we may do - like having EndpointShards with specific gateway. All this will come in master. 

So a 'grouping' of endpoints is EndpointShard - and may have a set of common properties, to avoid having to iterate/check individual endpoints.  To scale to 1M+ endpoints we need to do extra optimizations. 

Re. model.Service - same reason I didn't reuse NetworkEndpoint, it has too much stuff we don't need, since it combines everything. We need lighter structs, matching the protocol.

For now replacing NetworkEndpoint is the main goal - but eventually model.Service will need to be reworked too.